### PR TITLE
Search npm and some other fixes in the new arquitecture

### PR DESCRIPTION
**Pull request by sarvaje**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [x] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/contributing/pull-requests/#commitmessages)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [ ] Added/Updated related tests.

### CODE REVIEW

File: packages/rule-no-vulnerable-javascript-libraries/snyk-snapshot.json [hunk 1]
```diff
@@ -0,0 +1 @@
...fiers":{"CWE":["CWE-79"],"CVE":["CVE-2013-4941"],"NSP":332,"ALTERNATIVE":["SNYK-JS-YUI-10386"]},"severity":"medium","semver":{"unaffected":[">=3.10.0 || <3.0.0"],"vulnerable":["<3.10.0 >=3.0.0"]},"credit":["Aleksandr Dobkin"],"CVSSv3":"CVSS:3.0/AV:N/AC:L/PR:N/UI:R/S:U/C:L/I:L/A:N","disclosureTime":"2013-05-14T21:00:00.000Z","patches":[],"publicationTime":"2017-02-13T08:54:05.822Z","modificationTime":"2017-02-13T08:54:05.822Z","creationTime":"2017-01-22T08:54:05.822Z","id":"npm:yui:20130515","cvssScore":5.4,"alternativeIds":["SNYK-JS-YUI-10386"]},{"title":"Cross-site Scripting (XSS)","moduleName":"yui","packageName":"yui","language":"js","packageManager":"npm","identifiers":{"CWE":["CWE-79"],"CVE":["CVE-2013-4940"],"ALTERNATIVE":["SNYK-JS-YUI-10387"]},"severity":"medium","semver":{"unaffected":[">=3.10.3 <3.10.2"],"vulnerable":["=3.10.2"]},"credit":["Unknown"],"CVSSv3":"CVSS:3.0/AV:N/AC:L/PR:N/UI:R/S:U/C:L/I:L/A:N","disclosureTime":"2013-06-03T21:00:00.000Z","patches":[],"publicationTime":"2017-02-13T09:01:24.863Z","modificationTime":"2017-02-13T09:01:24.863Z","creationTime":"2017-01-22T09:01:24.863Z","id":"npm:yui:20130604","cvssScore":5.4,"alternativeIds":["SNYK-JS-YUI-10387"]}]}}
```
Comment (alrra) [inline packages/rule-no-vulnerable-javascript-libraries/snyk-snapshot.json]: This needs to be in `src/` so I'll remove it.

File: packages/sonarwhal/package.json [hunk 2]
```diff
@@ -112,20 +119,14 @@
   "license": "Apache-2.0",
   "main": "./dist/src/lib/sonarwhal.js",
   "name": "sonarwhal",
-  "nyc": {
```
Comment (alrra) [inline packages/sonarwhal/package.json]: Let's keep it here.

File: .gitignore [hunk 3]
```diff
@@ -10,3 +10,4 @@ dist
 node_modules
 npm-debug.log
 yarn-error.log
+**/.vscode/
```
Comment (alrra) [inline .gitignore]: Line 7 can  be chanced to `.vscode` and this line removed.

**Comment by alrra** [comment]
@sarvaje: I merged the part about updating the copy scripts in https://github.com/sonarwhal/sonarwhal/commit/f56c9d6933c602832d860c33374edfd86163d455.

Can you rebase and extract the chrome lunching related things in a separate commit, and the npm related things in another commits? Thanks!

**Comment by sarvaje** [comment]
>Can you rebase and extract the chrome lunching related things in a separate commit, and the npm related things in another commits? Thanks!

I will try hahahaha

**Comment by alrra** [comment]
> I will try hahahaha

If not, let me know and I will do it (I didn't do it because you better understand what is related to what change). Thanks!

**Comment by alrra** [comment]
> You will need to fix some things again

@sarvaje I actually took care of that. :)

File: packages/sonarwhal/src/lib/cli/init.ts [hunk 4]
```diff
@@ -19,10 +20,52 @@ import { debug as d } from '../utils/debug';
...
+        }
+
+        logger.log('Packages intalled successfully');
+    } catch (err) {
+        /*
+         * There was an error installing packages.
+         * Show message to install packages manually.
+         */
+        logger.error(err);
+        logger.log(`Something when wrong installing package, please run:
```
Comment (molant) [inline packages/sonarwhal/src/lib/cli/init.ts]: > Something went wrong installing the packages,

File: packages/sonarwhal/src/lib/cli/init.ts [hunk 5]
```diff
@@ -19,10 +20,52 @@ import { debug as d } from '../utils/debug';
...
+    for (const [key, value] of Object.entries(rules)) {
+        if (value !== 'off') {
+            packages.push(`@sonarwhal/rule-${key}`);
+        }
+    }
+
+    const command: string = `npm install ${packages.join(' ')}${global ? ' -g' : ''}`;
+
+    try {
+        logger.log(`Running command ${command}`);
```
Comment (molant) [inline packages/sonarwhal/src/lib/cli/init.ts]: Shouldn't this be a `debug`?

File: packages/sonarwhal/src/lib/cli/init.ts (hunk 4 above)
Comment (molant) [reply packages/sonarwhal/src/lib/cli/init.ts]: This `logger.log` should probably be a `logger.error`

File: packages/sonarwhal/src/lib/cli/init.ts [hunk 6]
```diff
@@ -19,10 +20,52 @@ import { debug as d } from '../utils/debug';
...
+
+        logger.log('Packages intalled successfully');
+    } catch (err) {
+        /*
+         * There was an error installing packages.
+         * Show message to install packages manually.
+         */
+        logger.error(err);
+        logger.log(`Something when wrong installing package, please run:
+${process.platform === 'linux' ? 'sudo ' : ''}${command}
```
Comment (molant) [inline packages/sonarwhal/src/lib/cli/init.ts]: What's the value for Mac platforms? Maybe this should be `process.platform !== 'windows'`

File: packages/sonarwhal/src/lib/cli/init.ts [hunk 7]
```diff
@@ -19,10 +20,52 @@ import { debug as d } from '../utils/debug';
...
+to install all the rules.`);
+    }
+};
+
 /** Initiates a wizard to gnerate a valid `.sonarwhalrc` file based on user responses. */
```
Comment (molant) [inline packages/sonarwhal/src/lib/cli/init.ts]: > Initiates the wizard to generate a valid

File: packages/sonarwhal/src/lib/cli/init.ts [hunk 8]
```diff
@@ -33,14 +76,13 @@ export const initSonarwhalrc = async (options: CLIOptions): Promise<boolean> =>
...
-    const rulesIds = resourceLoader.getCoreRules();
-    const rulesConfig = rulesIds.reduce((config, ruleId) => {
-        config[ruleId] = 'warning';
-
-        return config;
-    }, {});
-
-    const rules = resourceLoader.loadRules(rulesConfig);
+    const npmRules: Array<NpmPackage> = await resourceLoader.getCoreRulesFromNpm();
+    const rules = npmRules.map((rule) => {
```
Comment (molant) [inline packages/sonarwhal/src/lib/cli/init.ts]: We need to find a way to know which ones are recommended or not for the defaults.
E.g.: `amp-validator`, `cloudinary`, etc. shouldn't be on by default.

Do we have access to the keywords? Maybe we can use that to filter?

File: packages/sonarwhal/src/lib/types.ts [hunk 9]
```diff
@@ -74,3 +74,23 @@ export interface IORA {
     fail(): void;
     text: string;
 }
+
+/**
+ * Represent an user in a npm package.
```
Comment (molant) [inline packages/sonarwhal/src/lib/types.ts]: > A user in a npm package.

File: packages/sonarwhal/src/lib/types.ts [hunk 10]
```diff
@@ -74,3 +74,23 @@ export interface IORA {
...
+ */
+export type NpmMaintainer = {
+    email: string;
+    username: string;
+};
+
+/**
+ * Represent the result of a npm search.
```
Comment (molant) [inline packages/sonarwhal/src/lib/types.ts]: > The result of a npm search.

File: packages/sonarwhal/src/lib/utils/resource-loader.ts [hunk 11]
```diff
@@ -280,3 +284,65 @@ export const loadFormatter = (formatterId: string): IFormatter => {
...
+                resolve(results);
+            });
+    });
+};
+
+const loadNpm = () => {
+    return npmLoadAsync({ loaded: false });
+};
+
+export const getNpmPackages = async (searchTerm: string = 'sonarwhal'): Promise<Array<NpmPackage>> => {
```
Comment (molant) [inline packages/sonarwhal/src/lib/utils/resource-loader.ts]: Need som `jsdoc` here

File: packages/sonarwhal/src/lib/utils/resource-loader.ts [hunk 12]
```diff
@@ -280,3 +284,65 @@ export const loadFormatter = (formatterId: string): IFormatter => {
...
+    return searchNpmPackages(searchTerm);
+};
+
+const filterPackages = (packages: Array<NpmPackage>, initTerm: string) => {
+    return packages.filter((pkg) => {
+        return pkg.name.startsWith(initTerm);
+    });
+};
+
+export const getExternalRulesFromNpm = async () => {
```
Comment (molant) [inline packages/sonarwhal/src/lib/utils/resource-loader.ts]: Need some `jsdoc` here

File: packages/sonarwhal/src/lib/utils/resource-loader.ts [hunk 13]
```diff
@@ -280,3 +284,65 @@ export const loadFormatter = (formatterId: string): IFormatter => {
...
+    return npmLoadAsync({ loaded: false });
+};
+
+export const getNpmPackages = async (searchTerm: string = 'sonarwhal'): Promise<Array<NpmPackage>> => {
+    await loadNpm();
+
+    return searchNpmPackages(searchTerm);
+};
+
+const filterPackages = (packages: Array<NpmPackage>, initTerm: string) => {
```
Comment (molant) [inline packages/sonarwhal/src/lib/utils/resource-loader.ts]: Move this up so the `export`s are all together.

Also add description saying it filters by `startsWith`

File: packages/sonarwhal/src/lib/utils/resource-loader.ts [hunk 14]
```diff
@@ -280,3 +284,65 @@ export const loadFormatter = (formatterId: string): IFormatter => {
...
+    const rules = await getNpmPackages('sonarwhal-rule');
+
+    /*
+     * We need to filter the results because the search can
+     * include other packages that doesn't start with `sonarwhal-rule`.
+     */
+    return filterPackages(rules, 'sonarwhal-rule');
+};
+
+export const getCoreRulesFromNpm = async () => {
```
Comment (molant) [inline packages/sonarwhal/src/lib/utils/resource-loader.ts]: @alrra are we going to separate also the formatters, etc.? Right now?

If so then maybe we should have a `getCorePackages` that accepts a filter and build `getCore{Rule|Parser|Formatter|Connector}sFromNpm` on top of it?

Comment (alrra) [reply packages/sonarwhal/src/lib/utils/resource-loader.ts]: > are we going to separate also the formatters, etc.? Right now?

I think we should do it.

File: packages/sonarwhal/src/lib/cli/init.ts (hunk 4 above)
Comment (sarvaje) [reply packages/sonarwhal/src/lib/cli/init.ts]: I'm not sure. The previous `logger.error` is the one printing the error. This `logger.log` is just to tell the user how to install manually the dependencies.

File: packages/sonarwhal/src/lib/cli/init.ts (hunk 8 above)
Comment (sarvaje) [reply packages/sonarwhal/src/lib/cli/init.ts]: yep, we can read the keywords, which one do you want to use for that?

Comment (molant) [reply packages/sonarwhal/src/lib/cli/init.ts]: `recommended`, `sonarwhal-recommended` or something like that. @alrra?

File: packages/sonarwhal/src/lib/cli/init.ts (hunk 4 above)
Comment (molant) [reply packages/sonarwhal/src/lib/cli/init.ts]: But this message is because there has been an error, right? If so it should go to that output.

Comment (sarvaje) [reply packages/sonarwhal/src/lib/cli/init.ts]: ok

File: packages/sonarwhal/src/lib/cli/init.ts (hunk 8 above)
Comment (sarvaje) [reply packages/sonarwhal/src/lib/cli/init.ts]: @alrra @molant done!

Comment (molant) [reply packages/sonarwhal/src/lib/cli/init.ts]: For the record, we decided to go with `sonarwhal-recommended` so it's more scoped.

**Pull request merged**